tests/lazy_test.go

```diff
@@ -30,20 +29,20 @@ func lazyCanLazilyLoadEntity(t *testing.T, driver *RavenTestDriver) {
 	{
 		session := openSessionMust(t, store)
 		query := session.Advanced().Lazily()
-		lazyOrder := query.Load(reflect.TypeOf(&Company{}), "companies/1", nil)
+		var order *Company
+		lazyOrder := query.Load(&order, "companies/1", nil)
```

Comment:
I'm not sure whatever this is a good idea. This looks like I can start using `order` now, but it would still be null afteward, right?

Reply:
This is the best I was able to come up so far to make this strongly typed API.

In order to avoid a very awkward:
```
lazy := query.Load()
valueInterface, err := lazy.GetValue()
value := valueInterface.(*Company)
```

we want:
```
var c *Company
```

and then pass `&c` somewhere in the API. I went with `Load(&c, ...)` in this case because, given the current structure of the code, I was able to make it work.

Ideally, it would be:
```
lazy := query.Load()
var c *Company
err := lazy.GetResult(&c)
```

While this is code and therefore, in theory, everything is possible, making it work this way would require more restructuring. Basically, the computation that is currently done in `Load()` would have to be delayed until `Lazy.GetValue()`. And then do the same for every other API that returns `Lazy` value.

I can give it a go to see how feasible that would be. Let me know.

Reply:
How about:

```
var order *Company
lazyOrder := query.Load(&order, "companies/1", nil)
var c *Company
err := lazyOrder.GetValue(&c)
```

So we'll set both the original `order` and the `c` provided.

Another thing to worry about is whatever the `order` in the previous code is stack allocated?

Reply:
Re: stack allocation. Are you worried about correctness or performance? If correctness, Go compiler will make the right call and will promote local variable from stack to heap if it escapes the local scope. If performance: in all those examples the pointer lives on stack but the value it'll be assigned will be a heap value.

Regarding the API.

There are 2 things I'm not happy about this way. 

But first, to make sure we're on the same page about the mechanics. In:
```
var order *Company
lazyOrder := query.Load(&order, "companies/1", nil)
var c *Company
err := lazyOrder.GetValue(&c)
```
`order` is never touched, it's only passed so that we can get the type of the result, which is required to build `fromToken` before sending the command to the server.

That's misleading API. The user would have an expectation that passing a pointer to a value implies it'll be either read or set. That's the first thing. 

I would rather see it revert to sth. like:
```
query := LoadOfType(reflect.TypeOf(&Company{}, "companies/1", nil))
```
This is easier to make it work (but keep in mind that there are more APIs like this that would require similar treatment).

The second issue I see: we pass duplicate information. 

Passing `reflect.TypeOf(&Company{})` is redundant and introduces a possibility of a mismatch between the argument to `Load()` and argument to `GetResult()`.

This problem of redundancy is common to all the `Load` and `Query` APIs, both lazy and non-lazy.

I've been trying to find ways to avoid the need to pass duplicate information.  In case of Lazy APIs it does mean the need to pass the result earlier than ideal.

Reply:
Both suggestions are trivial to implement. See https://gist.github.com/kjk/4ff3bd8dce086757168f3ede507e518d for untested sketch.

It's your call.

My opinion is that of 2 bad solutions, the current version is less bad.

The only difference between the current code and this proposal is that `GetValue` either requires or optionally allows for providing information that has already been provided earlier by the user, which is confusing.

Reply:
Okay, let's leave it for now as is, then.

Reply:
Another option could be to rename `GetValue()` to `Execute()` and add `ExecuteAndGetValue(result interface{})` with understnading that `Execute()` would be the default and `ExecuteAndGetValue()` is for cases where where `GetValue()` is far apart from the place that creates the `Lazy` value (which is not the scenario exercised in test cases).